### ISSUE TITLE
Improvements/optimizations to instruction set

### ISSUE THREAD

**Issue by rrthomas** [rrthomas/smite#92]
Is 8-bit encoding OK?

Suggestion: we use 8-bit encoding, and add 1-byte literals as extra instructions.

**Comment by apt1002** [comment]
8-bit is okay, yes, but what's wrong with 6-bit? We have plenty of spare opcodes. The specializer ignores extra instructions because they are rare and irregular, so making the literals into extra instructions sounds wrong to me.

**Comment by rrthomas** [comment]
8-bit is much easier to encode and decode. But given `I` the impetus to use a specific number of bits is much less.

Surely the specializer just pays attention to certain instructions, so there's no problem with teaching it about specific extra instructions (there's nothing different about extra instructions other than their opcodes).

**Comment by apt1002** [comment]
In principle the specializer could understand extra instructions, yes.

How do you make the instruction/extra distinction, then? I thought it was about what is/not specified to exist as standard on every implementation. It also seemed to coincide with a threshold of expense and rarity. By all these criteria, one-byte literals are not extra.

**Comment by rrthomas** [comment]
I may be engaging in polypyloctemy here, but the intention is to keep the core tiny (I'm going for 1,000 lines of C), and have everything else on top. So yes, it's about expense and rarity, but also about simplicity.

Extra instructions, in particular, are not supposed to be differentiated from normal instructions except by not being implemented in the core. Most practical implementations will want I/O, so need libc extra instructions; ditto one-byte literals. I'd expect the current specialization scheme to handle extra instructions just like the core set without change, and indeed to be a useful cross-check (ooh, this extra instruction is being used much more often than this core instruction!).

"Extra but included by default" is how I'd characterize one-byte literals. This means the default build always tests the "extras" machinery.

However, none of this prevents one-byte literals (eventually) migrating (back to) the core.

**Comment by apt1002** [comment]
I support the goal of keeping SMite tiny.

I suppose I can imagine a version without 1-byte literals. I suppose the specializer ought to do its best in the presence of word-size literal zeros and ones (assuming that word-size literal values are included in instruction traces). But I'm not sure it's a sensible design.

We know small literals make up a large fraction of instructions executed. Judged by static or dynamic frequency, they deserve a bit of core opcode space; more so than e.g. division or XOR, which are rare and could be subroutines. It's a little crazy to make "SWAP" and "DUP" six bytes long (including the required literal) in the core instruction encoding.

I don't really buy the "extra but included by default" story. That just seems like a perverse definition of "extra". I suggest that core SMite should make sense as a product, e.g. for embedding, and that default SMite should add the development environment and platform integration. The requirement for 1-byte literals (and XOR and division) arises in core; the requirement for console and file I/O does not arise unless you have a console and a filing system.

And then there is the missed opportunity of making the instruction/extra boundary coincide with the fast path/slow path boundary. It is convenient and future-proof for the specializer to just ignore the extra instructions. Probably the same will be true in other applications.

In summary, I think 1-byte literals should be in core.

**Comment by rrthomas** [comment]
I don't understand the fast path/slow path bit. Extra instructions don't have to be any slower than core instructions. For some applications it might be useful to add extra instructions that are executed frequently (e.g. for crypto). I don't understand why the specializer cares, and doesn't simply act according to the frequencies of instructions that it measures.

**Comment by apt1002** [comment]
As it stands, extra instructions have to go through the main dispatch logic, then a call to `smite_extra()`, then a switch on the library, then another subroutine call, and then yet another switch, before they can be executed. The dispatch tree is appropriate, but not fast.

The typed arguments and results are an obstacle to optimisation too. Ordinary stack items can share stack cache variables of type `smite_WORD`, but the typed items can't. Currently, the specializer spills all the cached items to memory before calling `smite_extra()`.

Another reason to dump all the cached stack items to memory is that the StackPictures of the extra instructions are lies. The true stack effect of an extra instruction depends on the reason code in a way which StackPicture can't model.

None of these difficulties is a problem, because the `smite_extra()` call abstracts them all away. I want to keep it that way.

**Comment by apt1002** [comment]
We should discuss your crypto example.

**Comment by apt1002** [comment]
Encode NEXT, CALL, BRANCH, and possibly BRANCHZ with high bits clear, so they can be encoded in fewer than 6 bits, and therefore fit in the remainder of a word.

**Comment by rrthomas** [comment]
There are still some things I would like to do here:

- It is I think just too annoying to have a 6-bit encoding. I think we should revert to 8-bit. This detail doesn't matter at all when buried in a code generator or debugger, but it really messes up exploration, and in general raises the bar for interacting with SMite (e.g. look at the extra burden imposed on pForth, which doesn't have a problem with either ARM (single instructions that occupy a word) or Beetle (bytes packed into a word). In any case, you said this should be easy to change, so it could be a parameter (like `WORD_SIZE` and `ENDISM`, one that would result in incompatible binaries). I note that even with `WORD_SIZE`=4, pForth rarely fills a whole word of 6-bit instructions. (The plans below fit in 6 bit encoding with room for a few extra instructions.)
- I want to add immediate literals. At least, a few `LITx` instructions (still not sure which ones are the best, though my recurrent favourite is zero plus powers of 2 up to maximum `WORD_SIZE` (32)). However, it is bugging me that the minimum space required to encode a single byte is a word; bytes are common and so are 16-bit quantities. I would be happy, however, to handle these in the same way as on Beetle, i.e. the payload has to be in the same word.
- For the same reason, I want to add `LOAD` and `STORE` of powers of 2 numbers of bytes less than a word. Or possibly "up to 64 bits". Or "up to a double word" (and have some instructions that aren't implemented on smaller-than-maximum `WORD_SIZE`s). And/or lower the maximum `WORD_SIZE` to 16 or 8.
- I think Beetle was right: `BRANCH[Z]`, `CALL` and `LIT` should have "immediate" versions that offer an argument in the remainder of the instruction word. (Note that this still leaves the space after `HALT` to be used in future! As well as any other extra instructions declared terminal—which makes me think, maybe instructions should have a "terminal bit" in their spec?) This allows single-word subroutine threading (even on 4-byte `WORD_SIZE` it's 26-bit).
- Extra instructions should not be allowed to lie about their stack effect. So (re)introduce a single instruction, `TRAP` or `ESCAPE` or `SWI` that does arbitrary stuff and is used for library access and all other off-piste action. (@apt1002's "slow path".) No reason not to use the stack effect mechanism to implement them for now, though, as at present; but in future the FFI machinery will evolve.
- Restore the argument to `POP`.

**Comment by apt1002** [comment]
- 6-bit: I think pForth would perhaps fill words more often if its inliner were more aggressive. Writing SMite code by hand seems to fill words. I'm a little disappointed with 8-bit but I can work with it and I understand your reasons.

 - Immediate literals: You know I want some more literal instructions too. For use before SWAP and DUP, please include small literals such as 3 and 5 in addition to powers of two. I recall we measured (in an older version of SMite that had two stacks and ROT instructions) that the most common literals are 0 and 1, distantly followed by 2 and -1. Larger literals were mostly labels, which can be common in a particular binary, but not in a way that generalises enough to permit an optimisation of the instruction set. I can believe that LITW_PC_REL uses many sub-word-sized literals, but it would not be very SMitish to optimise them. I suggest that the best design is to allocate opcodes for small literals up to (say) 7, and otherwise use LITW (or LITW_PC_REL).

 - Sub-word-sized LOAD and STORE: no strong objection. These will certainly be used in compiled C code. Please include sign-extending versions of LOAD. Having said that, these operations can be implemented as subroutines fairly well, and are not that common, so it's not unreasonable to omit them.

 - Immediate versions of BRANCH etc: I'm not entirely convinced, but let me talk it through. Immediate CALL is the big win. If you have an immediate form of BRANCH, then the non-immediate form is just RET, which would certainly be needed anyway. Presumably immediate BRANCH and CALL would be PC-relative. Together they perhaps obviate LITW_PC_REL (the remaining rare cases can be handled using a subroutine). Non-immediate CALL can be implemented as a pair of subroutines (one PC-relative and one not) and does not warrant an opcode. I don't like immediate LIT because it is not terminal; better to segregate opcodes and literal data into separate words. I don't mind immediate BRANCHZ so much but I'm still against it. In summary, I think a suitable design would have LITW, BRANCHZ unchanged; LITW_PC_REL removed; immediate PC-relative BRANCH and CALL added; the existing BRANCH renamed to RET; and the existing CALL removed. Adding a terminal flag sounds sensible in any case.

 - EXT: I agree whole-heartedly.

 - Argument to POP: probably not worth it. I like the current level of generality of StackEffect and it would go beyond it.

**Comment by apt1002** [comment]
Oh, and I see no need for WORD_SIZE to exceed 8 this century.

**Comment by rrthomas** [comment]
Argument to POP: this just works, what's the problem? I implemented it and fixed up the tests, and the only code change is to `vm_data.py`:

    POP = Instruction(0x04, ['ITEMS:', 'COUNT'], [], '')

(No bogus cast required any longer either!)

I have pushed this to my PR so you can see it and try it, rather than as a _fait accompli_.

**Comment by rrthomas** [comment]
`WORD_SIZE`: I can see at least two reasons to allow it to get bigger.

The first is for bigger address spaces as in https://www.labs.hpe.com/the-machine ; or how about running networking code where you're dealing with IPv6 addresses all the time?

The second is crypto: the EVM, for example, uses 32-byte quantities all the time.

Both of these are specialised uses, and I don't see why SMite shouldn't have such specialized incarnations. However, as a matter of priority this is a question of leaving the door open where possible.

**Comment by rrthomas** [comment]
Variable-sized `LOAD`/`STORE`: my concern here is that these will often be used in a loop. If it were possible easily to recognise and inline a subroutine, or to convert an inlined sequence into a native instruction, that would be fine; but doing it after the fact seems rather perverse given how easy it is to add to the instruction set, in a way that doesn't really complicate the design (given that we already have `LOADB`/`STOREB`), just adds a few more cases.

**Comment by rrthomas** [comment]
Immediate operands for branches etc.: I can't fault your reasoning, but I'd rather not make the instruction set irregular, so simply adding immediate versions of `CALL`, `BRANCH` and `BRANCHZ` would be my preferred option (the latter makes generating code for `else` the same as `if` and for an unconditional loop the same as a conditional loop). Having allowed non-terminal immediate `BRANCHZ`, it seems reasonable to have non-terminal immediate `LIT`, and, keeping it regular, make its argument a word too, so that all the immediate literals are decoded identically (not the same as Beetle).

**Comment by rrthomas** [comment]
Small immediate literals: much to-ing and fro-ing in my brain, but I think either up to 8 (largest common scalar size), or up to `WORD_SIZE` (becauee `WORD_SIZE`-sized objects will typically be common) inclusive, and 8 seems a bit arbitrary if one accepts the possibility of larger word sizes.

**Comment by rrthomas** [comment]
6-bit: we can test 6 vs 8 easily. If 6 performs well then `INSTRUCTION_BITS` could at least be a register like `WORD_SIZE`, and we could see whether others agree with me (ease of hacking) or you (performance grr!).

**Comment by apt1002** [comment]
non-terminal immediate BRANCHZ <-- Eh? How can it be non-terminal if the bits after it hold an immediate constant?

Do you mean that BRANCHZ's immediate operand will be in the next whole word? That sounds sensible.

Edit: no that's clearly not what you mean, because you talk about a uniform encoding. Please clarify.

**Comment by apt1002** [comment]
Variable-sized `LOAD`/`STORE`: Yes, do it.

**Comment by apt1002** [comment]
Small immediate literals: the most interesting cases are stack positions, I think. Address offsets are the next most interesting cases, and are generally divisible by a power of two (I think you mean that WORD_SIZE is a common address offset).

We have to draw a line somewhere. We can offer a few small literals as dedicated opcodes without changing any part of the assembler, disassembler, decoder, specializer etc. The set of literal values could be 0 to 10 or it could be 0, 1, 2, 3, 4, 6, 8, 12, 16, 24, 32, or something else. I don't mind. That seems like a good place to draw the line. Supporting a wider variety of literals requires a new encoding and is probably not worth it.

**Comment by apt1002** [comment]
6-bit: Okay, let's do that.

**Comment by apt1002** [comment]
makes generating code for else the same as if <-- I don't understand this.

I think you are concerned about patching the targets of forward branches. Clearly assembling a BRANCHZ can't be the same as assembling a BRANCH, but perhaps patching the targets could be? Is that what you mean?

If that's your concern, it seems to be best addressed by a LITW followed by the non-immediate form of BRANCH or BRANCHZ. In particular, that's the only way to be sure that the target will be in the reachable range. It also means the patching operation is just a STORE.

**Comment by apt1002** [comment]
Immediate operands for branches etc.: I don't want to tell you what to do. I'm trying to infer your priorities and to suggest how to meet them.

You say dense code is a concern for you, and that you don't mind doing some work to achieve it. The low-hanging fruit for dense code is immediate CALL, as you observed. Maybe focus on that first? I am persuaded of its wisdom.

CALL doesn't need patching (correct?), so the instruction-set regularity argument does not apply. Don't worry about it.

You gave no use case for non-immediate (i.e. indirect) CALL. I can't see one, as I said. It can be achieved quite efficiently anyway by CALLing the following code: 0 SWAP BRANCH.

After that, if you want to make the code denser still, I would focus on writing a proper assembler. And after that we can see if there's still anything to gain from immediate forms of BRANCH and BRANCHZ.

Maybe there will be. However, if you do it the other way around (immediate BRANCH first, then a proper assembler) you will not be focusing on your priorities as well, and you will have complicated the design. That doesn't make sense, I think.

**Comment by apt1002** [comment]
Argument to POP: I stand corrected. I thought that there was a requirement for ITEMS to be in the same place afterwards, but I must have misunderstood.

**Comment by rrthomas** [comment]
POP: the documentation was ambiguous; I've improved it.

**Comment by rrthomas** [comment]
My priorities (as ever) are largely aesthetic and narrative. Asymmetries in a design provoke uncertainty and criticism, which requires effort to combat (I had a book on Modula-3 for my degree which had an appendix of "Just-so stories" on various features of the language). Maintaining symmetry when it's performance-neutral and adds little complexity (as with e.g. multi-size `LOAD`/`STORE`) is probably a gain in reduced explicative effort over time, and it looks good too.

**Comment by rrthomas** [comment]
Indirect call: most sorts of late binding would naturally use this, e.g. vtables.

I'm not really interested in making code dense as such, just not unnecessarily wasteful of space.

For reasonable range of addresses, patching the remainder of a word after a one-byte opcode is fine.

I think I had a different definition of "terminal" in my head from yours. I meant "execution does not proceed past this instruction by incrementing PC", whereas it seems you mean "no more opcodes in the current word".

Conclusion: I think we're close enough to agreement and what I have in mind is simple enough that I'll just do it and we then have something concrete to discuss.

**Comment by rrthomas** [comment]
Closing: changes are either done in incoming PRs or deferred to other issues.

**Comment by rrthomas** [comment]
See also #12.
